### PR TITLE
feat(bazel): simplify ng_package by dropping esm5 and fesm5

### PR DESCRIPTION
**Pull request by IgorMinar**
esm5 and fesm5 are no longer needed and have been deprecated in the past.

https://v9.angular.io/guide/deprecations#esm5-and-fesm5-code-formats-in-angular-npm-packages

This commit modifies ng_package to no longer distribute these two formats in npm packages
built by ng_package (e.g. @angular/core).

This commit intentionally doesn't fully clean up the ng_package rule to remove all traces of esm5 and fems5
build artifacts as that is a bigger cleanup and currently we are narrowing down the scope of this change
to the MVP needed for v10, which in this case is 'do not put esm5 and fesm5' into the npm packages.

More cleanup to follow: https://angular-team.atlassian.net/browse/FW-2143

### CODE REVIEW

**Comment by IgorMinar** [comment]
some of the integration tests are still failing, but the core part of the change is done and ready for the review.

**Comment by IgorMinar** [comment]
@devversion or @alan-agius4 if you have free cycles could you please help me resolve the last remaining test failures? I suspect that most of them are just fs path updates (fesm5->fesm2015) and golden file updates.

if you can't I'll keep on working on this file tomorrow morning.

**Comment by devversion** [comment]
@googlebot I consent.

**Comment by alan-agius4** [comment]
@IgorMinar, @devversion offered to take this up and try to make it green.

File: packages/bazel/src/ng_package/packager.ts [hunk 1]
```diff
@@ -342,6 +348,19 @@ function main(args: string[]): number {
...
+    // If we guessed the index paths for a module, and it contains an explicit `package.json`
+    // file that already sets format properties, we skip automatic insertion of format
+    // properties but report a warning in case properties have been set by accident.
+    if (moduleData.guessedPaths && !isGeneratedPackageJson &&
+        hasExplicitFormatProperties(parsedPackage)) {
+      console.error('WARNING: `package.json` explicitly sets format properties (like `main`).');
+      console.error(
+          '    Skipping automatic insertion of format properties as explicit ' +
+          'format properties are set.');
+      console.error('    Ignore this warning if explicit properties are set intentionally.');
```
Comment (IgorMinar) [inline packages/bazel/src/ng_package/packager.ts:360]: this DX is not great. Any time we print a warning that says "ignore this warning if ..." we failed. Could we instead say "add `customFormatResolutions = True` attribute to the target if you really want to provide custom format resolution in your npm package" - needs a bit more word-smithing :-)

**Review by IgorMinar — COMMENTED** [review]
otherwise @devversion's addition looks good to me.

File: packages/bazel/src/ng_package/packager.ts (hunk 1 above)
Comment (devversion) [reply packages/bazel/src/ng_package/packager.ts:360]: Agreed 👍 

We will have a follow-up for this as we have a similar warning already and it might make sense to think about these warnings in a combined sense. i.e. having the flag disable both of these warnings.

File: packages/bazel/test/ng_package/BUILD.bazel [hunk 2]
```diff
@@ -45,6 +49,10 @@ jasmine_node_test(
         "//packages/common:npm_package",
         "@npm//shelljs",
     ],
+    # The "ng_package" rule currently does not properly handle flat module
+    # bundles which the common package tests rely on. We temporarily disable
+    # this test until we fix `ng_package` for Ivy. Tracked with: FW-2144.
+    tags = ["fixme-ivy-aot"],
```
Comment (IgorMinar) [inline packages/bazel/test/ng_package/BUILD.bazel:55]: lgtm! thank

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts [hunk 3]
```diff
@@ -89,11 +116,11 @@ runInEachFileSystem(() => {
 
       expect(() => mainNgcc({
                basePath: '/dist',
-               propertiesToConsider: ['es2015', 'fesm5', 'module'],
+               propertiesToConsider: ['es2015', 'module'],
```
Comment (petebacondarwin) [inline packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:119]: why this change?

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts [hunk 4]
```diff
@@ -280,7 +311,7 @@ runInEachFileSystem(() => {
         `
       });
 
-      compileIntoFlatEs5Package('test-package', {
+      compileIntoFlatEs2015Package('test-package', {
```
Comment (petebacondarwin) [inline packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:314]: Why this change?
Could you not just change line 347 to be the following?

```
propertiesToConsider: ['esm5', 'esm2015'],
```

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts [hunk 5]
```diff
@@ -389,7 +420,7 @@ runInEachFileSystem(() => {
 
     // This is necessary to ensure XPipeDef.fac is defined when delegated from injectable def
     it('should always generate factory def (fac) before injectable def (prov)', () => {
-      compileIntoFlatEs5Package('test-package', {
+      compileIntoFlatEs2015Package('test-package', {
```
Comment (petebacondarwin) [inline packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:423]: Similarly here.

**Review by petebacondarwin — APPROVED** [review]
LGTM - couple of minor nits.

Reviewed for fw-ngcc

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts (hunk 3 above)
Comment (IgorMinar) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:119]: @devversion can you please answer?

I'm guessing that Paul saw that we removed fesm5 from everywhere and try to remove it from here. @petebacondarwin now that you mentioned this, it's likely that this change is a mistake and the line should be rolled back.

Comment (IgorMinar) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:119]: ... because existing libraries on npm still use `fesm5` and ngcc might want to consider them...

Comment (devversion) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:119]: Yep, good catch. I think since we don't rely on any of the actual release packages here, this change is unnecessary. Reverting!

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts (hunk 4 above)
Comment (devversion) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:314]: I was thinking that in APF primarily we ship ES2015 code now, and this test doesn't mention any es5 specifics, so I changed it to the flat `es2015` package creation.

Happy to revert if you think we should.

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts (hunk 5 above)
Comment (devversion) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:423]: Same response as above.

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:314]: I think it is safe to leave it as-is.

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts (hunk 3 above)
Comment (devversion) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:119]: Just to be clear: I did not remove `esm5` everywhere. This particular was just an oversight. I kept dedicated esm5 test as we could process libraries shipping with `<= APF 10`.

Primarily to ensure we still handle ES5 properly, I added a build target that downlevels the core `fesm2015` bundle to `fesm5` using Babel. That way we still get "integration test" coverage for ES5.

**Review by kara — APPROVED** [review]

**Review by kara — APPROVED** [review]
LGTM

Reviewed-for: global-approvers

**Review by alan-agius4 — APPROVED** [review]

**Review by devversion — APPROVED** [review]
LGTM for the changes I did not author.

**Comment by alan-agius4** [comment]
Super excited for this, farewell ES5 👋
